packages/edge-bundler/node/npm_dependencies.ts

```diff
@@ -105,101 +121,89 @@ interface GetNPMSpecifiersOptions {
+    bundle: false,
+    entryPoints: [file],
+    logLevel: 'silent',
+    platform: 'node',
+    write: false,
+  })
+
+  return compiled.outputFiles[0].text
+}
+
+async function parseImportsForFile(file: string, rootPath: string) {
+  const source = TYPESCRIPT_EXTENSIONS.has(path.extname(file))
+    ? await compileTypeScript(file)
+    : await fs.readFile(file, 'utf-8')
+
+  return await parseImports(source, {
+    resolveFrom: rootPath,
+  })
+}
+
 /**
  * Parses a set of functions and returns a list of specifiers that correspond
  * to npm modules.
  */
 const getNPMSpecifiers = async ({ basePath, functions, importMap, environment, rootPath }: GetNPMSpecifiersOptions) => {
   const baseURL = pathToFileURL(basePath)
-  const { reasons } = await nodeFileTrace(functions, {
-    base: rootPath,
-    processCwd: basePath,
-    readFile: async (filePath: string) => {
-      // If this is a TypeScript file, we need to compile in before we can
-      // parse it.
-      if (TYPESCRIPT_EXTENSIONS.has(path.extname(filePath))) {
-        const compiled = await build({
-          bundle: false,
-          entryPoints: [filePath],
-          logLevel: 'silent',
-          platform: 'node',
-          write: false,
-        })
-
-        return compiled.outputFiles[0].text
-      }
-
-      return fs.readFile(filePath, 'utf8')
-    },
-    resolve: async (specifier, ...args) => {
-      // Start by checking whether the specifier matches any import map defined
-      // by the user.
-      const { matched, resolvedImport } = resolve(specifier, importMap, baseURL)
-
-      // If it does, the resolved import is the specifier we'll evaluate going
-      // forward.
-      if (matched && resolvedImport.protocol === 'file:') {
-        const newSpecifier = fileURLToPath(resolvedImport).replace(/\\/g, '/')
-
-        return nftResolve(newSpecifier, ...args)
-      }
-
-      return nftResolve(specifier, ...args)
-    },
-  })
   const npmSpecifiers: { specifier: string; types?: string }[] = []
-  const npmSpecifiersWithExtraneousFiles = new Set<string>()
-
-  for (const [filePath, reason] of reasons.entries()) {
-    const parents = [...reason.parents]
-    const isExtraneousFile = reason.type.every((type) => type === 'asset')
-
-    // An extraneous file is a dependency that was traced by NFT and marked
-    // as not being statically imported. We can't process dynamic importing
-    // at runtime, so we gather the list of modules that may use these files
-    // so that we can warn users about this caveat.
-    if (isExtraneousFile) {
-      parents.forEach((path) => {
-        const specifier = getPackageName(path)
-
-        if (specifier) {
-          npmSpecifiersWithExtraneousFiles.add(specifier)
+  for (const func of functions) {
+    const imports = await parseImportsForFile(func, rootPath)
+
+    for (const i of imports) {
+      const specifier = i.moduleSpecifier.isConstant ? i.moduleSpecifier.value! : i.moduleSpecifier.code
```

Comment:
Why do we need the non-null assertion?

Reply:
I added a comment now about it, I'll paste it below also to save time going into the code for it:
>// The non-null assertion is required because typescript can not infer that `moduleSpecifier.value` can be narrowed to a string.
>// The narrowing is possible because `moduleSpecifier.value` will always be a string when `moduleSpecifier.isConstant` is true.